File: PE-common/src/main/java/nl/adaptivity/diagram/svg/SVGPath.java
/*
 * Copyright (c) 2016.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.diagram.svg;

import nl.adaptivity.diagram.DiagramPath;
import nl.adaptivity.diagram.Pen;
import nl.adaptivity.diagram.Rectangle;
import nl.adaptivity.lang.Doubles;

import java.util.ArrayList;
import java.util.List;


public class SVGPath implements DiagramPath<SVGPath>{

  private interface IPathElem {

    void appendPathSpecTo(StringBuilder builder);

    void getBounds(Rectangle storage, IPathElem previous, final Pen<?> stroke);

    double getX();
    double getY();
  }

  private static abstract class OperTo implements IPathElem {
    final double mX;
    final double mY;

    public OperTo(double x, double y) {
      mX = x;
      mY = y;
    }

    @Override
    public double getX() {
      return mX;
    }

    @Override
    public double getY() {
      return mY;
    }
  }

  private static class MoveTo extends OperTo {
    public MoveTo(double x, double y) {
      super(x, y);
    }

    @Override
    public void appendPathSpecTo(StringBuilder builder) {
      builder.append("M").append(mX).append(' ').append(mY).append(' ');
    }

    @Override
    public void getBounds(final Rectangle storage, final IPathElem previous, final Pen<?> stroke) {
      storage.set(mX, mY, 0d,0d);
    }
  }

  private static class LineTo extends OperTo {
    public LineTo(double x, double y) {
      super(x, y);
    }

    @Override
    public void appendPathSpecTo(StringBuilder builder) {
      builder.append("L").append(mX).append(' ').append(mY).append(' ');
    }

    @Override
    public void getBounds(final Rectangle storage, final IPathElem previous, final Pen<?> stroke) {
      final double sw = stroke.getStrokeWidth();
      if (Doubles.isFinite(sw) && sw > 0d) {
        double hsw = sw / 2;
        double height = getY() - previous.getY();
        double width = getX() - previous.getX();
        double angle = Math.atan(height / width);
        double extendX = Math.abs(Math.sin(angle) * hsw);
        double extendY = Math.abs(Math.cos(angle) * hsw);
        storage.set(Math.min(previous.getX(), mX)-extendX, Math.min(previous.getY(), mY)-extendY, Math.abs(previous.getX() - mX)+2*extendX, Math.abs(previous.getY() - mY)+2*extendY);
      } else {
        storage.set(Math.min(previous.getX(), mX), Math.min(previous.getY(), mY), Math.abs(previous.getX() - mX), Math.abs(previous.getY() - mY));
      }
    }
  }

  private static class CubicTo extends OperTo {
    private final double mCX1;
    private final double mCY1;
    private final double mCX2;
    private final double mCY2;

    public CubicTo(double cX1, double cY1, double cX2, double cY2, double x, double y) {
      super(x, y);
      mCX1 = cX1;
      mCY1 = cY1;
      mCX2 = cX2;
      mCY2 = cY2;
    }

    @Override
    public void appendPathSpecTo(StringBuilder builder) {
      builder.append("C").append(mCX1).append(' ').append(mCY1).append(' ')
                          .append(mCX2).append(' ').append(mCY2).append(' ')
                          .append(mX).append(' ').append(mY).append(' ');
    }

// Property accessors start
    @Override
    public void getBounds(final Rectangle storage, final IPathElem previous, final Pen<?> stroke) {
      final double hsw=stroke.getStrokeWidth()/2;
      final double previousX = previous.getX();
      final double previousY         = previous.getY();

      final double dx = previousX;
      final double cx = (3 * mCX1) - (3 * previousX);
      final double bx = ((3 * mCX2) - (6 * mCX1)) + (3 * previousX);
      final double ax = ((mX - (3 * mCX2)) + (3 * mCX1)) - previousX;

      final double dy = previousY;
      final double cy = (3 * mCY1) - (3 * previousY);
      final double by = ((3 * mCY2) - (6 * mCY1)) + (3 * previousY);
      final double ay = ((mY - (3 * mCY2)) + (3 * mCY1)) - previousY;

      final double dax = 3*(getX() - (3 * mCX2) + (3 * mCX1) - previousX);
      final double day = 3*(getY() - (3 * mCY2) + (3 * mCY1) - previousY);
      final double dbx = 2*(((3 * mCX2) - (6 * mCX1)) + (3 * previousX));
      final double dby = 2*(((3 * mCY2) - (6 * mCY1)) + (3 * previousY));
      final double dcx = (3 * mCX1) - (3 * previousX);
      final double dcy = (3 * mCY1) - (3 * previousY);


      double left = Math.min(previousX, mX);
      double right = Math.max(previousX, mX);
      double top = Math.min(previousY, mY);
      double bottom = Math.max(previousY, mY);

      double t1=0d;
      double t2=0d;
      double t3=0d;
      double t4=0d;

      if (dax!=0d) {
        t1 = (-dbx + Math.sqrt((dbx * dbx) - (4 * dax * dcx))) / (2 * dax);
        t2 = (-dbx - Math.sqrt((dbx * dbx) - (4 * dax * dcx))) / (2 * dax);
      }

      if (day!=0d) {
        t3 = (-dby + Math.sqrt((dby * dby) - (4 * day * dcy))) / (2 * day);
        t4 = (-dby - Math.sqrt((dby * dby) - (4 * day * dcy))) / (2 * day);
      }

      for(double t: new double[]{t1, t2, t3, t4}) {
        if ((t > 0d) && (t < 1d)) {
          final double x = (((((ax * t) + bx) * t) + cx) * t) + dx;
          left = Math.min(x, left);
          right = Math.max(x, right);

          final double y = (((((ay * t) + by) * t) + cy) * t) + dy;
          top = Math.min(y, top);
          bottom = Math.max(y, bottom);
        }
      }

      storage.set(left-hsw, top-hsw, right - left + stroke.getStrokeWidth(), bottom - top + stroke.getStrokeWidth());
      
    }
// Property acccessors end

  }

  private static class Close implements IPathElem {

    @Override
    public void appendPathSpecTo(StringBuilder builder) {
      builder.append("Z ");
    }

    @Override
    public void getBounds(final Rectangle storage, final IPathElem previous, final Pen<?> stroke) {
      storage.set(previous.getX(), previous.getY(), 0d, 0d);
    }

    @Override
    public double getX() {
      return Double.NaN;
    }

    @Override
    public double getY() {
      return Double.NaN;
    }
  }

  private List<IPathElem> mPath = new ArrayList<>();

  @Override
  public SVGPath moveTo(double x, double y) {
    mPath.add(new MoveTo(x, y));
    return this;
  }

  @Override
  public SVGPath lineTo(double x, double y) {
    mPath.add(new LineTo(x, y));
    return this;
  }

  @Override
  public SVGPath cubicTo(double x1, double y1, double x2, double y2, double x3, double y3) {
    mPath.add(new CubicTo(x1, y1, x2, y2, x3, y3));
    return this;
  }

  @Override
  public SVGPath close() {
    mPath.add(new Close());
    return this;
  }

  public String toPathData() {
    StringBuilder result = new StringBuilder();
    for(IPathElem elem: mPath) {
      elem.appendPathSpecTo(result);
    }
    return result.toString();
  }

  @Override
  public Rectangle getBounds(final Rectangle dest, final Pen<?> stroke) {
    if (mPath.size()==1) {
      final IPathElem elem = mPath.get(0);
      dest.set(elem.getX()-stroke.getStrokeWidth()/2, elem.getY()-stroke.getStrokeWidth()/2, stroke.getStrokeWidth(), stroke.getStrokeWidth());
    } else if (mPath.size()>1) {
      Rectangle tmpRect = new Rectangle(0d,0d,0d,0d);
      mPath.get(1).getBounds(dest, mPath.get(0), stroke);
      for (int i = 2; i < mPath.size(); i++) {
        mPath.get(i).getBounds(tmpRect, mPath.get(i-1), stroke);
        dest.extendBounds(tmpRect);
      }
    }
    return dest;
  }
}

File: PE-common/src/commonMain/kotlin/nl/adaptivity/diagram/svg/SVGPath.kt
/*
 * Copyright (c) 2018.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.diagram.svg

import nl.adaptivity.diagram.DiagramPath
import nl.adaptivity.diagram.Pen
import nl.adaptivity.diagram.Rectangle
import nl.adaptivity.util.multiplatform.append
import kotlin.math.*


class SVGPath : DiagramPath<SVGPath> {

    private val path = mutableListOf<IPathElem>()

    private interface IPathElem {

        val x: Double
        val y: Double

        fun appendPathSpecTo(builder: Appendable)

        fun getBounds(storage: Rectangle, previous: IPathElem, stroke: Pen<*>?)
    }

    private abstract class OperTo(override val x: Double, override val y: Double) : IPathElem

    private class MoveTo(x: Double, y: Double) : OperTo(x, y) {

        override fun appendPathSpecTo(builder: Appendable) {
            builder.append("M").append(x).append(' ').append(y).append(' ')
        }

        override fun getBounds(storage: Rectangle, previous: IPathElem, stroke: Pen<*>?) {
            storage[x, y, 0.0] = 0.0
        }
    }

    private class LineTo(x: Double, y: Double) : OperTo(x, y) {

        override fun appendPathSpecTo(builder: Appendable) {
            builder.append("L").append(x).append(' ').append(y).append(' ')
        }

        override fun getBounds(storage: Rectangle, previous: IPathElem, stroke: Pen<*>?) {
            // TODO this is not valid as it does not consider miters, nor does it support link closing
            val sw = stroke?.strokeWidth ?: 0.0
            if (sw.isFinite() && sw > 0.0) {
                val hsw = sw / 2
                val height = y - previous.y
                val width = x - previous.x
                val angle = atan(height / width)
                val extendX = abs(sin(angle) * hsw)
                val extendY = abs(cos(angle) * hsw)
                storage[min(previous.x, x) - extendX, min(previous.y, y) - extendY,
                    abs(previous.x - x) + 2 * extendX] = abs(previous.y - y) + 2 * extendY
            } else {
                storage[min(previous.x, x), min(previous.y, y), abs(previous.x - x)] =
                    abs(previous.y - y)
            }
        }
    }

    private class CubicTo(
        private val mCX1: Double,
        private val mCY1: Double,
        private val mCX2: Double,
        private val mCY2: Double,
        x: Double,
        y: Double
                         ) : OperTo(x, y) {

        override fun appendPathSpecTo(builder: Appendable) {
            builder.append("C").append(mCX1).append(' ').append(mCY1).append(' ')
                .append(mCX2).append(' ').append(mCY2).append(' ')
                .append(x).append(' ').append(y).append(' ')
        }


        override fun getBounds(storage: Rectangle, previous: IPathElem, stroke: Pen<*>?) {
            val hsw = stroke?.run { strokeWidth / 2 } ?: 0.0
            val previousX = previous.x
            val previousY = previous.y

            val cx = 3 * mCX1 - 3 * previousX
            val bx = 3 * mCX2 - 6 * mCX1 + 3 * previousX
            val ax = x - 3 * mCX2 + 3 * mCX1 - previousX

            val cy = 3 * mCY1 - 3 * previousY
            val by = 3 * mCY2 - 6 * mCY1 + 3 * previousY
            val ay = y - 3 * mCY2 + 3 * mCY1 - previousY

            val dax = 3 * (x - 3 * mCX2 + 3 * mCX1 - previousX)
            val day = 3 * (y - 3 * mCY2 + 3 * mCY1 - previousY)
            val dbx = 2 * (3 * mCX2 - 6 * mCX1 + 3 * previousX)
            val dby = 2 * (3 * mCY2 - 6 * mCY1 + 3 * previousY)
            val dcx = 3 * mCX1 - 3 * previousX
            val dcy = 3 * mCY1 - 3 * previousY


            var left = min(previousX, x)
            var right = max(previousX, x)
            var top = min(previousY, y)
            var bottom = max(previousY, y)

            var t1 = 0.0
            var t2 = 0.0
            var t3 = 0.0
            var t4 = 0.0

            if (dax != 0.0) {
                t1 = (-dbx + sqrt(dbx * dbx - 4.0 * dax * dcx)) / (2 * dax)
                t2 = (-dbx - sqrt(dbx * dbx - 4.0 * dax * dcx)) / (2 * dax)
            }

            if (day != 0.0) {
                t3 = (-dby + sqrt(dby * dby - 4.0 * day * dcy)) / (2 * day)
                t4 = (-dby - sqrt(dby * dby - 4.0 * day * dcy)) / (2 * day)
            }

            for (t in doubleArrayOf(t1, t2, t3, t4)) {
                if (t > 0.0 && t < 1.0) {
                    val x = ((ax * t + bx) * t + cx) * t + previousX
                    left = min(x, left)
                    right = max(x, right)

                    val y = ((ay * t + by) * t + cy) * t + previousY
                    top = min(y, top)
                    bottom = max(y, bottom)
                }
            }

            storage[left - hsw, top - hsw, right - left + hsw * 2] = bottom - top + hsw * 2

        }


    }

    private class Close : IPathElem {

        override val x: Double
            get() = Double.NaN

        override val y: Double
            get() = Double.NaN

        override fun appendPathSpecTo(builder: Appendable) {
            builder.append("Z ")
        }

        override fun getBounds(storage: Rectangle, previous: IPathElem, stroke: Pen<*>?) {
            storage[previous.x, previous.y, 0.0] = 0.0
        }
    }

    override fun moveTo(x: Double, y: Double): SVGPath {
        path.add(MoveTo(x, y))
        return this
    }

    override fun lineTo(x: Double, y: Double): SVGPath {
        path.add(LineTo(x, y))
        return this
    }

    override fun cubicTo(x1: Double, y1: Double, x2: Double, y2: Double, x3: Double, y3: Double): SVGPath {
        path.add(CubicTo(x1, y1, x2, y2, x3, y3))
        return this
    }

    override fun close(): SVGPath {
        path.add(Close())
        return this
    }

    fun toPathData(): String {
        return buildString {
            path.forEach { it.appendPathSpecTo(this) }
        }
    }

    override fun getBounds(dest: Rectangle, stroke: Pen<*>?): Rectangle {
        if (path.size == 1) {
            val elem = path.get(0)
            val strokeWidth = stroke?.strokeWidth ?: 0.0
            dest[elem.x - strokeWidth / 2, elem.y - strokeWidth / 2, strokeWidth] = strokeWidth
        } else if (path.size > 1) {
            val tmpRect = Rectangle(0.0, 0.0, 0.0, 0.0)
            path.get(1).getBounds(dest, path.get(0), stroke)
            for (i in 2 until path.size) {
                path.get(i).getBounds(tmpRect, path.get(i - 1), stroke)
                dest.extendBounds(tmpRect)
            }
        }
        return dest
    }
}
